### PR TITLE
[rule] Allow deleting attributes from Starlark `rule()` function

### PR DESCRIPTION
**Pull request by Yannic**
This is useful for rules like `rules_license`'s `license()` rule that are typically used via `package(default_package_metadata = [":license"]).

Since there's an implicit dependency from every target in the package to what's declared in `default_package_metadata`, rules intended for this purpose need to be wrapped in a macro today.

```starlark
_license = rule(
  # ...
)

def license(**kwargs):
  _license(package_metadata = [], **kwargs)
```

With this change, wrapping the rule is no longer necessary.

```starlark
license = rule(
  # ...
  deleted_attrs = ["package_metadata"],
)
```

### CODE REVIEW

**Comment by Yannic** [comment]
Converting to draft for now since this is purely a POC whether something like this would work at all.

@lberki do you see any obvious issues with exposing something like this to Starlark before I send time creating a design doc?

**Comment by fmeum** [comment]
The macro approach could be "inlined" via the new `initializer` attribute of `rule`. Maybe that's sufficient and a new attribute isn't needed?

**Comment by lberki** [comment]
Not on the implementation side. In fact, `RuleClass.Builder.removeAttribute()` is used quite a few times in built-in rules. So as far as the implementation goes, it's sound.  I think the bigger issue is how to use this: presumably, you'll want to delegate to the implementation of the rule you inherit from in some way, but that implementation will expect the attribute you delete.

I do understand what you want to use this for in `license()`, but it seems to be a pretty niche use case and if it really is a niche, it's probably better off implemented as a macro as you propose above.

**Comment by Yannic** [comment]
> The macro approach could be "inlined" via the new `initializer` attribute of `rule`. Maybe that's sufficient and a new attribute isn't needed?

Mhm I tried that yesterday, but my naive approach of filtering out `["package_metadata", "applicable_licenses"]` didn't work, so it seems like propagating attributes from `package()` to the target happens later than the initializer.. Let me try again with setting them to `[]` explicitly and report back.

Do I understand correctly that `initializer` is available in `8.x`, but not in `7.x`? If so, is this something we could potentially backport?

**Comment by Yannic** [comment]
Nope `initializer` doesn't work: `Error in license: Initializer can only set Starlark defined attributes, not 'package_metadata'`

```starlark
def _set_package_metadata(**kwargs):
    args = {k: v for k, v in kwargs.items()}
    args["package_metadata"] = []

    return args
```

**Comment by Yannic** [comment]
I can see that this feature becomes a foot-gun when extending an existing rule via subrules. Maybe we should only allow it only when there`s no `parent`?

I don't have a usecase outside on `license()` or package metadata in mind right now. We are working on making this more extendible right now to account for things other than licenses or allow organizations to override / extend what's available in the bzlmod dependency, so I do expect there to be more rules with a need to remove `package_metadata`. And having this would allow `rules_license` (or its successor we plan to create in the next week) to create an abstract "parent" rule with the attributes removed to make it easy for users (like we would if it was a native rule)

**Comment by hvadehra** [comment]
Maybe we can just allow `package_metadata` at https://github.com/bazelbuild/bazel/blob/11c0d7bbf22b1fc1bbe260af46c134802e73bb55/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleClassFunctions.java#L1846 instead?

**Comment by meisterT** [comment]
We are closing this draft PR given the feedback from @lberki that exposing general attribute deletion in `rule()` introduces subtle footguns for rule implementations and is better served by macros or targeted handling for `package_metadata`. 
  
Since this was opened as an exploratory proof-of-concept and has developed merge conflicts, closing it helps keep the backlog clean. If you'd like to revisit how `package_metadata` interacts with license rules in the future, please feel free to start a discussion or issue.

**Review by ortstadtthomas91-coder — APPROVED** [review]
